Implemented login page.

rottencandy opened the pull request:
Created a login page in mithril.js as part of the task 101 of the project.

rottencandy committed:
Add login.js

login.js when used as a module works as a mithril component that renders a
login page and communicates with RS Json API to faciliate validation.
Also updated qmake file to include login.js in build.

rottencandy committed:
Update rswebui.js to use modifiable API keys

Added extra function to verify API keys(to be used by login.js) and
updated HTTP requests function to use a different callback on failiure.

rottencandy committed:
Update index.html to use render login.js on start.

index.html's initial startpoint now loads all pages using mithril's
route functionality.
And changed downloads.js to disable updating of downloads list before the
keys have been verified.

rottencandy committed:
Update mithril from v0.2.3 to v1.1.6

rottencandy committed:
Modify api request to not run if keys unverified

Communication with Json API now only happens if the keys are verified.
Automatic rendering and reloading of downloads is handled by oninit attribute
of the downloads component.

G10h4ck commented:
so from my tests the login system you implemented seems to work, but probably involuntarily the download page is now broken and the download file list is never requested, about mithril update, I think we should not include it minified in our repository, it is impossible to review WTF has been really included, and if we need minified it should be done at compile time, not included in our repository, as it would pose serious security concerns about the webui...

G10h4ck commented:
So I have been looking at the differences between mithril.js minified and not minified, for such a  minimalist framework I think using the minified source is not worth the effort/risk, so we should remove the minified version completely, and make sure the RS webui keeps working

rottencandy commented:
Yes, the minified version is present, but it was not being used in the build process. I will look into the downloads page.

But I think debugging is still difficult, because what qmake essentially does is, take all the .js scripts and append them to a single file. Makes it hard to figure out from which script the error occurs.

rottencandy committed:
Remove minified mithril script

Due to security and debugging issues, the minified version will not
be used in the build process anymore. The framework is pretty minimalistic
so it would not make that big of a difference.

rottencandy committed:
Load downloads only after login is verified.

Previously, the downloads page would be loaded but API would not
be accessed. This was causing issues with mithril's auto rendering.
Not loading the downloads component will make sure no requests are
being sent before login finishes.

rottencandy commented:
Removed minified script.
I think the problem with downloads component was that it sends requests to the API server as soon as it is created (before login verification, thus failing). Now modified and should be created only after the login credentials have been verified.
